central/cloudsources/manager/manager.go

```diff
@@ -214,6 +238,80 @@ func (m *managerImpl) matchDiscoveredClusters(clusters []*discoveredclusters.Dis
 	}
 }
 
+func (m *managerImpl) changeStatusForDiscoveredClusters(clusterID string, status storage.DiscoveredCluster_Status) {
+	cluster, err := m.getCluster(clusterID)
+	if err != nil {
+		log.Errorw("Failed to get cluster to mark as secured",
+			logging.ClusterID(clusterID), logging.Err(err))
+		return
+	}
+
+	// If the cluster has no metadata, we can short-circuit since there won't be anything to update.
+	if cluster.GetStatus().GetProviderMetadata().GetCluster().GetId() == "" {
+		return
+	}
+
+	cloudSources, err := m.cloudSourcesDataStore.ListCloudSources(cloudSourceCtx, search.EmptyQuery())
+	if err != nil {
+		log.Errorw("Failed to list stored cloud sources for marking cluster as secured",
+			logging.Err(err), logging.ClusterID(clusterID))
+		return
+	}
+
+	discoveredClusters := m.fetchDiscoveredClusters(cluster, cloudSources)
+	// In case we found no discovered clusters, we can short-circuit here.
+	if len(discoveredClusters) == 0 {
+		return
+	}
+
+	unspecifiedProviders := concurrency.WithLock1(&m.unspecifiedProviderLock, func() set.StringSet {
```

Comment:
In general I think this is a possible TOC-TOU, right? I wonder if we need to do a stricter synchronization with the manager.

Reply:
So yeah its true that we are currently in the midst of matching and also update the secured cluster here. I don't think it's reasonable to hold a lock for the duration of the matching.

I'm also not really sure we require stricter synchronization between the two: it's going to be handled eventually consistent either way and the correct status will be updated in the next matching interval, not sure whether people will care too much about the consistency here.

Alternatively, we can drop the unspecified part here and let this be dealt with by the matching part in an eventually consistent manner as well. I'd be OK with both ways, I think the possible TOC-TOU gives more consistency than just leaving it unsecured, but WDYT?

Reply:
We've decided on removing the logic completely w.r.t the unspecified types, hence this isn't applicable anymore.